### PR TITLE
release-23.1: go.mod: cherry-pick etcd-io/raft#32

### PR DESCRIPTION
**Pull request by pav-kv**
This dependency upgrade includes a patched PR 32 from etcd-io/raft which fixes an out-of-memory scenario during election with large unapplied Raft logs.

Part of #104402
Epic: none
Release note (bug fix): Fixed a bug in upstream etcd-io/raft which could result in pulling unlimited amount of log into memory, and lead to out-of-memory situations. With the fix, the log scan has a limited memory footprint.
Release justification: fixing a bug after a customer escalation

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
The code looks good, but the [commit](https://github.com/cockroachdb/raft/commit/b160e656b5ae) is not from our fork's release-23.1 branch. Could you switch to that branch, cherry-pick the commit, and then bump to the resulting SHA? That way there won't be danger of someone putting the next cherry-pick on the correct branch and updating, which would _drop_ this fix again.

**Comment by pavelkalinnikov** [comment]
@tbg I was intending to bump the `crdb-release-23.1` SHA **after** merging in this PR (or at a point when it's clear that I can be merged). This way the branch won't be pointing at an "maybe broken" state optimistically.

> That way there won't be danger of someone putting the next cherry-pick on the correct branch and updating, which would drop this fix again.

This shouldn't happen (at least blindly). If we have concurrent dep upgrades, both will update the SHA in `go.mod` and conflict, so one will need to be rebased. While doing so, hopefully the rebasing person will realize that the base SHA has been updated, so they need to rebase the cherry-pick in the fork accordingly.

> Could you switch to that branch, cherry-pick the commit, and then bump to the resulting SHA?

The alternative "optimistic" approach that you're suggesting has the risk of: we update the SHA for the first cherry-pick, the other person cherry-picks another thing on top of it; but we realize that the first cherry-pick is broken. It's too late to update the `crdb-release-23.1` SHA now.

**Comment by tbg** [comment]
> This shouldn't happen (at least blindly). If we have concurrent dep upgrades, both will update the SHA in go.mod

Not concurrent - back to back.

I think you're right though, I was missing that release-23.1 is a fast-forward to [cherry-pick-pr-32](https://github.com/cockroachdb/raft/compare/cherry-pick-pr-32), so LGTM and please push after the merge.

**Review by tbg — APPROVED** [review]

**Pull request merged**
